### PR TITLE
Option to run CLI with a recipe file instead of command line args 

### PR DESCRIPTION
**Pull request by stuartlynn**
Adds a new sub command to the CLI 

```popgetter_cli recipe --recipe-file recipe.json --output-file result.geojson --output-format geojson```

Which loads in the DataRequestSpec from json and runs the request 

a sample request file: https://github.com/Urban-Analytics-Technology-Platform/popgetter-cli/blob/run_with_config/test_recipe.json

Some thoughts / questions 

- It might be worth the CLI having a top level struct that pairs a DataRequestSpec with an OutputOptions so that both can be read from the config file. We still want DataRequestSpec to be it's own thing as this is how python and JS will interact with the library without any need to specify output 
- We are starting with json as the config language but it should be possible to use toml or yaml or anything else that serde supports. Not sure if we have opinions on the best formats to use here.
- We can add some serde directives to make the json format a little easier to write, for example using different [enum representations](https://serde.rs/enum-representations.html) if they make sense

### CODE REVIEW

File: src/cli.rs [hunk 1]
```diff
@@ -141,6 +146,33 @@ impl RunCommand for SurveysCommand {
...
 }
 
+/// The Recipe command loads a recipy file and generates the output data requested
+#[derive(Args, Debug)]
+pub struct RecipeCommand{
+    #[arg(short, long)]
+    recipe_file: String,
+
+    #[arg(short='f', long)]
+    output_format: OutputFormat,
```
Comment (sgreenbury) [inline src/cli.rs:156]: Add information on the available output formats for `--help`

File: src/cli.rs [hunk 2]
```diff
@@ -141,6 +146,33 @@ impl RunCommand for SurveysCommand {
     }
 }
 
+/// The Recipe command loads a recipy file and generates the output data requested
+#[derive(Args, Debug)]
+pub struct RecipeCommand{
+    #[arg(short, long)]
+    recipe_file: String,
```
Comment (sgreenbury) [inline src/cli.rs:153]: Make `recipe_file` positional argument

File: src/cli.rs [hunk 3]
```diff
@@ -8,7 +8,7 @@ use popgetter::{
     formatters::{CSVFormatter, GeoJSONFormatter, GeoJSONSeqFormatter, OutputFormatter, OutputGenerator}, Popgetter
 };
 use serde::{Deserialize, Serialize};
-use std::{fs::File, str::FromStr};
+use std::{fs::{self, File}, str::FromStr};
```
Comment (sgreenbury) [inline src/cli.rs:11]: See if we can ignore warning as required

File: src/cli.rs [hunk 4]
```diff
@@ -141,6 +146,33 @@ impl RunCommand for SurveysCommand {
     }
 }
 
+/// The Recipe command loads a recipy file and generates the output data requested
+#[derive(Args, Debug)]
+pub struct RecipeCommand{
+    #[arg(short, long)]
+    recipe_file: String,
+
+    #[arg(short='f', long)]
```
Comment (sgreenbury) [inline src/cli.rs:155]: If output format not specified, try to infer from filename suffix

File: src/cli.rs [hunk 5]
```diff
@@ -68,10 +64,19 @@ impl RunCommand for DataCommand {
...
+    }
+}
+
+impl RunCommand for DataCommand {
+    async fn run(&self) -> Result<()> {
+        let popgetter = Popgetter::new()?;
+        let data_request = DataRequestSpec::from(self);
+        let mut results = popgetter.get_data_request(&data_request).await?;
 
         println!("{results:#?}");
```
Comment (sgreenbury) [inline src/cli.rs:77]: Replace this with env_logger: https://docs.rs/env_logger/latest/env_logger/# (but do this in separate PR)

File: src/cli.rs [hunk 6]
```diff
@@ -141,6 +146,33 @@ impl RunCommand for SurveysCommand {
...
+    output_file: String
+}
+
+impl RunCommand for RecipeCommand{
+    async fn run(&self) -> Result<()> {
+        let popgetter = Popgetter::new()?;
+        let config = fs::read_to_string(&self.recipe_file)?;
+        let data_request: DataRequestSpec = serde_json::from_str(&config)?;
+        let mut results = popgetter.get_data_request(&data_request).await?;
+        println!("{results}");
```
Comment (sgreenbury) [inline src/cli.rs:168]: Logging but for next PR

**Comment by sgreenbury** [comment]
Looks great! As discussed I'll open an issue that replaces prints with env logging.

File: src/cli.rs (hunk 6 above)
Comment (penelopeysm) [reply src/cli.rs:168]: Done in #36

File: src/cli.rs (hunk 5 above)
Comment (penelopeysm) [reply src/cli.rs:77]: Done in #36

File: src/cli.rs (hunk 1 above)
Comment (sgreenbury) [reply src/cli.rs:156]: Added elsewhere!

File: src/cli.rs (hunk 2 above)
Comment (sgreenbury) [reply src/cli.rs:153]: Done

File: src/cli.rs (hunk 3 above)
Comment (sgreenbury) [reply src/cli.rs:11]: No solution for now

File: src/cli.rs (hunk 4 above)
Comment (sgreenbury) [reply src/cli.rs:155]: Open as a new ticket #39

File: src/cli.rs [hunk 7]
```diff
@@ -51,13 +51,9 @@ pub struct DataCommand {
...
 }
 
-impl RunCommand for DataCommand {
-    async fn run(&self) -> Result<()> {
-        let popgetter = Popgetter::new()?;
-        let data_request = DataRequestSpec::from(self);
-        let mut results = popgetter.get_data_request(&data_request).await?;
-
-        let formatter = match &self.output_format{
+impl From<&OutputFormat> for OutputFormatter{
```
[the remark this answers is not in the document]
Comment (sgreenbury) [reply src/cli.rs:54]: Added
